### PR TITLE
Add steady state heat flux termination criteria

### PR DESCRIPTION
**Pull request by jdannberg**
This pull request add a termination criterion based on the steady state heat flux. 

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: include/aspect/termination_criteria/steady_heat_flux.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*
+  Copyright (C) 2019 by the authors of the ASPECT code.
```
Comment (bangerth) [inline include/aspect/termination_criteria/steady_heat_flux.h:2]:
```suggestion
  Copyright (C) 2021 by the authors of the ASPECT code.
```

Comment (naliboff) [inline include/aspect/termination_criteria/steady_heat_flux.h:2]:
```suggestion
  Copyright (C) 2021 by the authors of the ASPECT code.
```

File: include/aspect/termination_criteria/steady_heat_flux.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        /**
+         * Read the parameters this class declares from the parameter file.
+         */
+        void
+        parse_parameters (ParameterHandler &prm) override;
+
+      private:
+        double                                  necessary_time_in_steady_state;
+        double                                  allowed_relative_deviation;
+
```
Comment (naliboff) [inline include/aspect/termination_criteria/steady_heat_flux.h:80]: Add brief documentation for the above two parameters?

File: source/termination_criteria/steady_heat_flux.cc [hunk 3]
```diff
@@ -0,0 +1,212 @@
+/*
+  Copyright (C) 2019 - 2020 by the authors of the ASPECT code.
```
Comment (naliboff) [inline source/termination_criteria/steady_heat_flux.cc:2]:
```suggestion
  Copyright (C) 2021  by the authors of the ASPECT code.
```

File: source/termination_criteria/steady_heat_flux.cc [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  namespace TerminationCriteria
+  {
+    namespace internal
+    {
+      void trim_time_heat_flux_list (const double necessary_time_in_steady_state,
+                                     std::list<std::pair<double, double> > &time_heat_flux_list)
+      {
+        // Remove old times until we're at the correct time period
+        // but ensure at least two entries remain in the list (one old, one current timestep)
+        auto it = time_heat_flux_list.begin();
```
Comment (naliboff) [inline source/termination_criteria/steady_heat_flux.cc:35]: Out of curiosity, does `it` refer to "iterator time"?

File: include/aspect/termination_criteria/steady_heat_flux.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+         */
+        std::set<types::boundary_id> boundary_indicators;
+
+        /**
+         * A list of pairs (time, heat flux) that we have computed at
+         * previous time steps. This is used to determine when we have reached
+         * steady state.
+         */
+        std::list<std::pair<double, double> >   time_heat_flux;
+
```
Comment (bangerth) [inline include/aspect/termination_criteria/steady_heat_flux.h:93]:
```suggestion
```

File: include/aspect/termination_criteria/steady_heat_flux.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    namespace internal
+    {
+      /**
+       * A function that trims the handed over list and removes all entries from the front that are
+       * further back in time measured from the last entry than given by the first argument.
+       * Additionally it makes sure to always keep two entries in the list, if the list had
+       * two or more entries. Otherwise the function does not change the list.
+       */
+      void trim_time_heat_flux_list (const double necessary_time_in_steady_state,
+                                     std::list<std::pair<double, double> > &time_heat_flux_list);
```
Comment (bangerth) [inline include/aspect/termination_criteria/steady_heat_flux.h:42]: This function should just live in the `.cc` file. See #4133.

File: source/termination_criteria/steady_heat_flux.cc (hunk 3 above)
Comment (bangerth) [inline source/termination_criteria/steady_heat_flux.cc:2]:
```suggestion
  Copyright (C) 2021 by the authors of the ASPECT code.
```

File: source/termination_criteria/steady_heat_flux.cc [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+#include <aspect/termination_criteria/steady_heat_flux.h>
+#include <aspect/postprocess/heat_flux_map.h>
+
+namespace aspect
+{
+  namespace TerminationCriteria
+  {
+    namespace internal
+    {
+      void trim_time_heat_flux_list (const double necessary_time_in_steady_state,
```
Comment (bangerth) [inline source/termination_criteria/steady_heat_flux.cc:30]: just move the documentation of the function here and make the namespace anonymous

File: source/termination_criteria/steady_heat_flux.cc [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+        time_heat_flux_list.erase(time_heat_flux_list.begin(), it);
+      }
+    }
+
+    template <int dim>
+    bool
+    SteadyHeatFlux<dim>::execute()
+    {
+      std::vector<std::vector<std::pair<double, double> > > heat_flux_and_area =
```
Comment (bangerth) [inline source/termination_criteria/steady_heat_flux.cc:48]:
```suggestion
      const std::vector<std::vector<std::pair<double, double> > > heat_flux_and_area =
```

File: source/termination_criteria/steady_heat_flux.cc [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+      // If the length of the simulation time covered in the list is shorter than the
+      // specified parameter, we must continue the simulation
+      if ((time_heat_flux.size() <= 2)
+          ||
+          (time_heat_flux.back().first - time_heat_flux.front().first < necessary_time_in_steady_state))
+        return false;
+
+      // Remove old entries outside of current time window
+      internal::trim_time_heat_flux_list(necessary_time_in_steady_state,time_heat_flux);
```
Comment (bangerth) [inline source/termination_criteria/steady_heat_flux.cc:78]:
```suggestion
      trim_time_heat_flux_list(necessary_time_in_steady_state,time_heat_flux);
```

File: source/termination_criteria/steady_heat_flux.cc [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+          flux_min = std::min(flux_min, (*it).second);
+          flux_max = std::max(flux_max, (*it).second);
+          flux_sum += (((*it).second + flux_prev)/2.0)*((*it).first-time_prev);
+          time_prev = (*it).first;
+          flux_prev = (*it).second;
+        }
+
+      flux_mean = flux_sum/(time_heat_flux.back().first-time_heat_flux.front().first);
+
+      // If the min and max are within the acceptable deviation of the mean,
```
Comment (naliboff) [inline source/termination_criteria/steady_heat_flux.cc:96]: A general question about the approximations being made here: if `time_heat_flux` only contains two entries, that would mean the steady state heat flux is being computed from adjacent time steps. 

Is this reasonable? Or would this only happen on the second time step? 

I'm trying to remember how I used to determine when steady was reached and I recall using quite a few time steps. Again, likely a mute point.

**Review by bangerth — COMMENTED** [review]
Just small stuff.

File: source/termination_criteria/steady_heat_flux.cc [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+      // we are in steady state and return true, otherwise return false
+      deviation_max = std::max(flux_mean - flux_min, flux_max - flux_mean);
+
+      AssertThrow(std::abs(flux_mean) > std::numeric_limits<double>::min(),
+                  ExcMessage("The average heat flux is close to 0.0. The "
+                             "'steady state heat flux' plugin can not compute a "
+                             "relative deviation of heat flux in this case."));
+
+      if (deviation_max/flux_mean > allowed_relative_deviation)
+        return false;
```
Comment (naliboff) [inline source/termination_criteria/steady_heat_flux.cc:106]: Just to be sure, `return true` will not be reached if the above condition is met (e.g., even though it is not within the conditional statement)?

**Review by naliboff — COMMENTED** [review]
@jdannberg - Nice! A few comments and suggestions after a quick look, but I'm sure others will weigh in as well.

**Comment by naliboff** [comment]
Ah, @bangerth beat me to it 😄 (and also provide a thorough review 😉 )

**Comment by bangerth** [comment]
Two reviews is better than one. And you asked the more substantive questions -- I just checked what is different from the existing plugin :-)

File: source/termination_criteria/steady_heat_flux.cc (hunk 4 above)
Comment (jdannberg) [reply source/termination_criteria/steady_heat_flux.cc:35]: I think it just stands for iterator, but I only copied this from the steady state temperature plugin.

File: source/termination_criteria/steady_heat_flux.cc (hunk 10 above)
Comment (jdannberg) [reply source/termination_criteria/steady_heat_flux.cc:96]: Yes, that's right. But I would say, in that case, the user should pick a different "Time in steady state" in the input file, or set a minimum time step that is smaller than the "Time in steady state". If the time steps are substantially smaller than that, there should always be enough entries in the `time_heat_flux` vector.

File: source/termination_criteria/steady_heat_flux.cc (hunk 11 above)
Comment (jdannberg) [reply source/termination_criteria/steady_heat_flux.cc:106]: As soon as the `return false` is reached, the function will return the value false and no other code in the function will be executed. Was that the question?
If you want I can change this to 
```
      if (deviation_max/flux_mean > allowed_relative_deviation)
        return false;
      else
        return true;
```
I didn't think about what would be easier to read, I just copied it from an existing plugin.

**Comment by jdannberg** [comment]
Okay, thank you! I think I've addressed all comments.

**Review by bangerth — COMMENTED** [review]
OK from my side, but I'll let @naliboff see whether he's satisfied as well.

**Review by bangerth — APPROVED** [review]

File: source/termination_criteria/steady_heat_flux.cc (hunk 11 above)
Comment (naliboff) [reply source/termination_criteria/steady_heat_flux.cc:114]: No fine as is, just wanted to make sure I understood. Thanks for the explanation!

File: source/termination_criteria/steady_heat_flux.cc (hunk 10 above)
Comment (naliboff) [reply source/termination_criteria/steady_heat_flux.cc:104]: @jdannberg - Makes sense, thanks for the explanation! Could you add some of the above text either at the top of the test case prm file and /or in the documentation for the "Time in steady state" parameter? You could almost copy this verbatim into the prm file. Other than that, ready to be merged!

Comment (jdannberg) [reply source/termination_criteria/steady_heat_flux.cc:104]: I have added a few lines to the documentation of the input parameter. I thought that would be the place where it is most useful for the user.

**Comment by jdannberg** [comment]
@naliboff Can you have another look?

File: source/termination_criteria/steady_heat_flux.cc [hunk 12]
```diff
@@ -135,6 +135,12 @@ namespace aspect
                              Patterns::Double (0.),
                              "The minimum length of simulation time that the system "
                              "should be in steady state before termination. "
+                             "Note that if the time step size is similar to or larger than "
+                             "this value, the termination criterion will only have very few "
+                             "(in the most extreme case, just two) heat flux values to check. "
+                             "To ensure that a larger number of time steps are included in "
+                             "the check for steady state, this value should be much larger "
+                             "than the time step size. "
```
Comment (naliboff) [inline source/termination_criteria/steady_heat_flux.cc:143]: @jdannberg - Perfect, thank you!

**Review by naliboff — APPROVED** [review]
Looks great, thank you for the additional documentation and adding this feature!

**Pull request merged**